dali/util/dynlink_cuda.cc

```diff
@@ -0,0 +1,602 @@
+        GET_PROC(cuMemcpyAtoHAsync);
+        GET_PROC(cuMemcpy2DAsync);
+        GET_PROC(cuMemcpy3DAsync);
+        GET_PROC(cuMemsetD8);
+        GET_PROC(cuMemsetD16);
+        GET_PROC(cuMemsetD32);
+        GET_PROC(cuMemsetD2D8);
+        GET_PROC(cuMemsetD2D16);
+        GET_PROC(cuMemsetD2D32);
+        GET_PROC(cuArrayCreate);
+        GET_PROC(cuArrayGetDescriptor);
+        GET_PROC(cuArray3DCreate);
+        GET_PROC(cuArray3DGetDescriptor);
+        GET_PROC(cuTexRefSetAddress);
+        GET_PROC(cuTexRefSetAddress2D);
+        GET_PROC(cuTexRefGetAddress);
+    }
+
+    // The following functions are specific to CUDA versions
+    if (driverVer >= 4000)
+    {
+        GET_PROC(cuCtxSetCurrent);
+        GET_PROC(cuCtxGetCurrent);
+        GET_PROC(cuMemHostRegister);
+        GET_PROC(cuMemHostUnregister);
+        GET_PROC(cuMemcpy);
+        GET_PROC(cuMemcpyPeer);
+        GET_PROC(cuLaunchKernel);
+        GET_PROC(cuProfilerStop);
+    }
+
+    if (driverVer >= 3010)
+    {
+        GET_PROC(cuModuleGetSurfRef);
+        GET_PROC(cuSurfRefSetArray);
+        GET_PROC(cuSurfRefGetArray);
+        GET_PROC(cuCtxSetLimit);
+        GET_PROC(cuCtxGetLimit);
+    }
+
+    if (driverVer >= 3000)
+    {
+        GET_PROC(cuMemcpyDtoDAsync);
+        GET_PROC(cuFuncSetCacheConfig);
+#ifdef CUDA_INIT_D3D11
+        GET_PROC(cuD3D11GetDevice);
+        GET_PROC(cuD3D11CtxCreate);
+        GET_PROC(cuGraphicsD3D11RegisterResource);
+#endif
+        GET_PROC(cuGraphicsUnregisterResource);
+        GET_PROC(cuGraphicsSubResourceGetMappedArray);
+
+        if (cudaVersion >= 3020 && __CUDA_API_VERSION >= 3020)
+        {
+            GET_PROC_V2(cuGraphicsResourceGetMappedPointer);
+        }
+        else
+        {
+            GET_PROC(cuGraphicsResourceGetMappedPointer);
+        }
+
+        GET_PROC(cuGraphicsResourceSetMapFlags);
+        GET_PROC(cuGraphicsMapResources);
+        GET_PROC(cuGraphicsUnmapResources);
+        GET_PROC(cuGetExportTable);
+    }
+
+    if (driverVer >= 2030)
+    {
+        GET_PROC(cuMemHostGetFlags);
+#ifdef CUDA_INIT_D3D10
+        GET_PROC(cuD3D10GetDevice);
+        GET_PROC(cuD3D10CtxCreate);
+        GET_PROC(cuGraphicsD3D10RegisterResource);
+#endif
+#ifdef CUDA_INIT_OPENGL
+        GET_PROC(cuGraphicsGLRegisterBuffer);
+        GET_PROC(cuGraphicsGLRegisterImage);
+#endif
+    }
+
+    if (driverVer >= 2010)
+    {
+        GET_PROC(cuModuleLoadDataEx);
+        GET_PROC(cuModuleLoadFatBinary);
+#ifdef CUDA_INIT_OPENGL
+        GET_PROC(cuGLCtxCreate);
+        GET_PROC(cuGraphicsGLRegisterBuffer);
+        GET_PROC(cuGraphicsGLRegisterImage);
+#endif
+#ifdef CUDA_INIT_D3D9
+        GET_PROC(cuD3D9GetDevice);
+        GET_PROC(cuD3D9CtxCreate);
+        GET_PROC(cuGraphicsD3D9RegisterResource);
+#endif
+    }
+
+    return CUDA_SUCCESS;
+}
+
+bool cuInitChecked() {
+    static std::mutex m;
+    static bool initialized = false;
+
+    std::unique_lock<std::mutex> l(m);
```

Comment:
I think you can move this to line 599, because if it is initialized you don't have to guard it anyway.
Other issue I see now it that during initialization `res == CUDA_SUCCESS` is returned, but in case of failure only first call will return false, while other one will return true. Maybe we should preserve `res == CUDA_SUCCESS` as status and return it for consecutive calls instead of `true`?

Reply:
I thought about this, but I was not sure if we want DALI_ENFORCE two files two times

Reply:
Ok maybe this would be better:
```
bool cuInitChecked() {
    static std::mutex m;
    static bool initialized = false;

    if (initialized)
        return true;

    std::unique_lock<std::mutex> l(m);

    static CUresult res = cuInit(0, __CUDA_API_VERSION);
    initialized = (res == CUDA_SUCCESS);
    return initialized;
}
```

Reply:
This way we will keep trying to initialize - I don't know if after first failure it is possible to succeed later.

Reply:
I think that if it fails, we should just throw and destroy the pipeline, the user have to resolve why the CUDA driver couldn't be found or read